src/main/java/io/r2dbc/proxy/callback/CallbackSupport.java

```diff
@@ -0,0 +1,290 @@
+                    executionInfo.setResult(resultObj);
+
+                    // apply a function to flux-chain right after the original publisher operations
+                    if (onMap != null) {
+                        return onMap.apply(resultObj, executionInfo);
+                    }
+                    return resultObj;
+                })
+                .doOnComplete(() -> {
+                    // apply a consumer to flux-chain right after the original publisher operations
+                    // this is the first chained doOnComplete on the result publisher
+                    if (onComplete != null) {
+                        onComplete.accept(executionInfo);
+                    }
+                })
+                .doOnError(throwable -> {
+                    executionInfo.setThrown(throwable);
+                })
+                .doFinally(signalType -> {
+
+                    Instant startTime = startTimeHolder.get();
+                    Instant currentTime = this.clock.instant();
+
+                    Duration executionDuration = Duration.between(startTime, currentTime);
+                    executionInfo.setExecuteDuration(executionDuration);
+
+                    String threadName = Thread.currentThread().getName();
+                    long threadId = Thread.currentThread().getId();
+                    executionInfo.setThreadName(threadName);
+                    executionInfo.setThreadId(threadId);
+
+                    executionInfo.setProxyEventType(ProxyEventType.AFTER_METHOD);
+
+                    listener.onMethodExecution(executionInfo);
+                });
+
+
+        } else {
+            // for method that generates non-publisher, execution happens when it is invoked.
+
+            String threadName = Thread.currentThread().getName();
+            long threadId = Thread.currentThread().getId();
+            executionInfo.setThreadName(threadName);
+            executionInfo.setThreadId(threadId);
+
+            // invoke before method
+            executionInfo.setProxyEventType(ProxyEventType.BEFORE_METHOD);
+            listener.onMethodExecution(executionInfo);
+
+            Instant startTime = this.clock.instant();
+
+            Object result = null;
+            Throwable thrown = null;
+            try {
+                result = method.invoke(target, args);
+            } catch (InvocationTargetException ex) {
+                thrown = ex.getTargetException();
+                throw thrown;
+            } finally {
+                executionInfo.setResult(result);
+                executionInfo.setThrown(thrown);
+
+                Instant currentTime = this.clock.instant();
+                Duration executionDuration = Duration.between(startTime, currentTime);
+                executionInfo.setExecuteDuration(executionDuration);
+
+                executionInfo.setProxyEventType(ProxyEventType.AFTER_METHOD);
+                listener.onMethodExecution(executionInfo);
+            }
```

Comment:
Other ways I can think of is to create a StopWatch, or make a template for the pattern with Consumer.

Can you expand a little bit about try-resource object?

Reply:
Essentially wrap things in a templated method. Perhaps Consumer or Supplier is a best fit. Try-Resource is something that must eventually be closed like a file, so maybe my comment was a little off.

Reply:
Tried to template the flow, but since it involves with Flux operations, not quite readable than directly putting logic on each Flux methods.
Instead, just made a simple `StopWatch` utility class, and delegated the `Duration` calculation, which seems cleaner approach.